session.go

```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
 	sess.breakCh = c
 }
 
-func (sess *session) handleRequests(reqs <-chan *gossh.Request) {
-	for req := range reqs {
-		switch req.Type {
-		case "shell", "exec":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+func (sess *session) handleRequests(ctx Context, reqs <-chan *gossh.Request) {
+	keepAliveEnabled := sess.keepAliveInterval > 0
+	lastReceived := time.Now()
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.rawCmd = payload.Value
+	var keepAliveCallback func()
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
-			}
+	var keepAliveTicker *time.Ticker
+	if keepAliveEnabled {
+		keepAliveTicker = time.NewTicker(sess.keepAliveInterval)
+		defer keepAliveTicker.Stop()
 
-			sess.handled = true
-			req.Reply(true, nil)
+		keepAliveCallback = func() {
+			lastReceived = time.Now()
+			keepAliveTicker.Reset(sess.keepAliveInterval)
+		}
 
-			go func() {
-				sess.handler(sess)
-				sess.Exit(0)
-			}()
-		case "subsystem":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+		ctx.SetValue(ContextKeyKeepAliveCallback, keepAliveCallback)
+	} else {
+		// Configure a stopped ticker to prevent `<-keepAliveTicker.C` from panicking.
+		keepAliveTicker = time.NewTicker(math.MaxInt64)
+		keepAliveTicker.Stop()
+	}
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.subsystem = payload.Value
+	for {
+		select {
+		case <-keepAliveTicker.C:
+			if lastReceived.Add(time.Duration(sess.keepAliveCountMax) * sess.keepAliveInterval).Before(time.Now()) {
+				log.Println("Keep-alive reply not received. Close down the session.")
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
+				err := sess.Close()
+				if err != nil {
+					log.Printf("Closing session failed: %v", err)
+				}
+				return
 			}
 
-			handler := sess.subsystemHandlers[payload.Value]
-			if handler == nil {
-				handler = sess.subsystemHandlers["default"]
+			log.Println("Send keep-alive request to the client")
```

Comment:
Have you confirmed logging is safe in these code paths and won't garble SSH sessions? (For instance, `ssh -vvv ...`.)

Reply:
`ssh -vvv ....` is the client command, and we're printing logs on the server side, right? Please clarify as maybe I'm missing something.

Reply:
You're right that this specific code is probably not problematic (except for maybe verbose logs on the agent?), but what about e.g. the handler? I'm assuming the client needs to handle keep-alives too? OpenSSH obviously does, but what about `coder ssh` in non-stdio mode?

Reply:
> I'm assuming the client needs to handle keep-alives too? OpenSSH obviously does, but what about coder ssh in non-stdio mode?

I don't know if default Go SSH client supports keep-alives, but if not, then most likely we will have to adjust the non-stdio code. I intend to focus on this once I'm done with changes in the library.

Reply:
I limited prints.